packages/hash/api/src/graphql/resolvers/taskExecutor/index.ts

```diff
@@ -15,10 +24,185 @@ export const executeDemoTask: Resolver<
+  [k: string]: unknown;
+}>;
+
+const streamNameToEntityTypeName = (name: string) => {
+  const sanitizedName = singular(upperFirst(camelCase(name)));
+  return `Github${sanitizedName}`;
+};
+
+export const executeGithubDiscoverTask: Resolver<
+  Promise<string>,
+  {},
+  LoggedInGraphQLContext,
+  MutationExecuteGithubDiscoverTaskArgs
+> = async (
+  _,
+  { config },
+  { dataSources: { db, taskExecutor }, user, logger },
+) => {
+  if (!taskExecutor) {
+    throw new ApolloError(
+      "A task-executor wasn't started, so external tasks can't be started",
+    );
+  } else {
+    try {
+      const catalog: AirbyteCatalog = await taskExecutor.runTask(
+        Tasks.GithubDiscover,
+        config,
+      );
+
+      const existingEntityChecker = await CachedEntityTypes(db, user);
+
+      for (const message of catalog) {
+        if (!message.name || !message.json_schema) {
+          continue;
+        }
+
+        const entityTypeName = streamNameToEntityTypeName(message.name);
+
+        if (existingEntityChecker.getExisting(entityTypeName) === undefined) {
+          await existingEntityChecker.createNew(
+            entityTypeName,
+            message.json_schema,
+          );
+          logger.debug(`Created a new EntityType: ${entityTypeName}`);
+        }
+      }
+
+      return JSON.stringify(catalog);
+    } catch (err: any) {
+      throw new ApolloError(`Task-execution failed: ${err}`);
+    }
+  }
+};
+
+export const executeGithubReadTask: Resolver<
+  Promise<string>,
+  {},
+  LoggedInGraphQLContext,
+  MutationExecuteGithubReadTaskArgs
+> = async (
+  _,
+  { config },
+  { dataSources: { db, taskExecutor }, user, logger },
+) => {
+  if (!taskExecutor) {
+    throw new ApolloError(
+      "A task-executor wasn't started, so external tasks can't be started",
+    );
+  } else {
+    const createdEntities = [];
+    try {
+      const airbyteRecords: AirbyteRecords = await taskExecutor.runTask(
+        Tasks.GithubRead,
+        config,
+      );
+      logger.debug(`Received ${airbyteRecords.length} records from Github`);
+
+      const existingEntityChecker = await CachedEntityTypes(db, user);
+      for (const record of airbyteRecords) {
+        const entityTypeName = streamNameToEntityTypeName(record.stream);
+        /** @todo - Check if entity already exists */
+        const entityTypeId = existingEntityChecker.getExisting(entityTypeName);
+        if (!entityTypeId) {
+          throw new Error(
+            `Couldn't find EntityType for ingested data with name: ${entityTypeName}`,
+          );
+        }
+
+        /** @todo - check primary key to see if entity already exists */
+        // Insert the entity
+        const entity = await Entity.create(db, {
+          accountId: user.accountId,
+          createdByAccountId: user.accountId,
+          versioned: true,
+          entityTypeId,
+          properties: record.data,
+        });
+
+        createdEntities.push(entity.entityId);
+      }
```

Comment:
I'm a little worried about doing this. I'm not sure if this will overwhelm the DB right now (I would need to look up the implementation of promises and see if they're batched at all or if it ends up dependent on available threads and such but I suspect because it's IO that it will spam with possibly 1000s of requests in one go). Also I don't think the existingEntityChecker is set up for concurrency right now and without putting a little thought into the caching mechanism I'd not be comfortable adding this.

Don't get me wrong, the suggestion to parallelise is 100% valid, but I think we're solving it in the wrong place and way here. We need to think about the API better, especially around entityType discovery and mass data-insertion. We'll address that anyway soon so I don't think we need a specific asana task around it (and this code will be revisited and rewritten)

Reply:
Sounds good